Fix accumulator and bias types coherence for int16x8 FC/Conv2d/TransposeConv2d operators

SaoirseARM opened the pull request:
This PR implements a fix for this issue: https://github.com/tensorflow/tensorflow/issues/53763

This adds a quantized_bias_type parameter to the Conv2d, TransposeConv2d and FullyConnected operators in the flatbuffer schema. 

If _experimental_full_integer_quantization_bias_type is set in the TfLiteConverter, quantized_bias_type will be set to this. 
If not, the bias->type will default int64 for int16x8.

Best regards,

Saoirse

SaoirseARM committed:
Adding quantized_bias_type to Flatbuffers schema to ensure bias->type is set/defaults correctly

SaoirseARM committed:
Added tests for Transpose_conv2d and Conv2d

gbaned commented:
@SaoirseARM Can you please resolve conflicts? Thank you!

SaoirseARM committed:
Merge remote-tracking branch 'origin/master' into toupstream/int6x8_32_accum

SaoirseARM commented:
> @SaoirseARM Can you please resolve conflicts? Thank you!

Thanks! conflicts should be resolved. Best regards, Saoirse

SaoirseARM committed:
Merge remote-tracking branch 'origin/master' into toupstream/int6x8_32_accum

thaink commented on tensorflow/lite/tools/optimize/quantize_model.cc:
```diff
@@ -272,6 +272,39 @@ std::unordered_set<string> PopulateRealValueOpSet(
+          options->quantized_bias_type = bias_type;
+        }
+      }
+      if (op_code && op_code->builtin_code == BuiltinOperator_CONV_2D) {
```
Nit: "If else" or switch.

SaoirseARM replied:
Thanks. I have updated this.

thaink commented on tensorflow/lite/tools/versioning/op_version.cc:
```diff
@@ -155,6 +168,19 @@ int GetBuiltinOperatorVersion(const OpSignature& op_sig) {
+          op_sig.outputs.at(0).type == kTfLiteInt16) {
+        // `quantized_bias_type` is supported at version 10.
+        if (fully_connected_params->quantized_bias_type) {
+          return 10;
```
Where is op version 9?

SaoirseARM replied:
op version 9 seems to be here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/versioning/op_version.cc#L204 
Please note, I had to update the op version to 11 as it seems op version 10 has been added. 
Please let me know if I should not do this. Thanks.

thaink replied:
SGTM

thaink commented on tensorflow/lite/schema/schema.fbs:
```diff
@@ -566,6 +566,9 @@ table Conv2DOptions {
   dilation_w_factor:int = 1;
   dilation_h_factor:int = 1;
+  // Parameters for Conv2D version 7 or above.
+  // If set, use this for bias accumulator.
```
"bias and accumulator" here and otheres.

SaoirseARM replied:
Thanks. Updated.

thaink commented on tensorflow/lite/kernels/transpose_conv.cc:
```diff
@@ -558,9 +572,29 @@ void EvalQuantizedPerChannel16x8(
                             weights->params.zero_point ||
                             output->params.zero_point;
 
-  // Fallback to reference kernel when bias_type is int64 as
```
Why removing these comments?

SaoirseARM replied:
I think these comments are no longer relevant with these changes. I am happy to add these back in though.

thaink replied:
I think the fact that we doesn't have optimized kernel for int64 bias doesn't change after this cl.

Please add these comment at line #598.

thaink commented on tensorflow/lite/kernels/conv.cc:
```diff
@@ -817,34 +829,35 @@ void EvalQuantizedPerChannel16x8(TfLiteContext* context, TfLiteNode* node,
                             filter->params.zero_point ||
                             output->params.zero_point;
 
-  // Fallback to reference kernel when bias_type is int64 as
```
Why remove these comments?

thaink reviewed (commented):
Could you also add a comment in tensorflow/lite/python/lite.py that the accumulator type will be the same as bias type set by full_integer_quantization_bias_type?

SaoirseARM committed:
Updates to int16x8 bias and accumulator

thaink commented:
Generally look good to me. Only a nit that I think the comments in transpose_conv.cc and conv.cc should be kept (just move to the right place).

After that I'll happy to give an approval from MOT side.

SaoirseARM committed:
Add extra comments to conv.cc and transpose_conv.cc

SaoirseARM commented:
> Generally look good to me. Only a nit that I think the comments in transpose_conv.cc and conv.cc should be kept (just move to the right place).
> 
> After that I'll happy to give an approval from MOT side.

thank you. I have added the comments back in. 
Best regards,
Saoirse

SaoirseARM committed:
Merge remote-tracking branch 'origin/master' into toupstream/int6x8_32_accum

gbaned commented:
@SaoirseARM Can you please resolve conflicts? Thank you!

rewu93 commented:
Repeat my question in #53763 here : do we really want accumulators to be 32bits? 16 bits activation X 8 bits weights FC means that the maximum dimension we can have before integer overflow is 2^8 = 256.

Tessil committed:
Check that the quantized_bias_type has been set before assignement.

The quantized_bias_type parameter default to 'FLOAT32 = 0' TensorType when not set in the FlatBuffer. Check if the read quantized_bias_type is not the default value.

Tessil committed:
Assign scratch_buffer->type using the quantized_bias_type from OpData instead of TfLiteTransposeConvParams

Tessil committed:
Only call the optimzed quantized FullyConnectedPerChannel kernel when quantized_bias_type is kTfLiteInt32 as it only support int32 bias

Tessil commented:
We are currently setting the default value of `quantized_bias_type` to `TensorType::FLOAT32 = 0` in the FlatBuffer which will be the default value for older models generated without the new `quantized_bias_type` variable. Ideally we would like to have something like `NONE` as default but it isn't there. 

The rationale to not use `TensorType::INT64` as default type is that there may be one day where int8 kernels could eventually support smaller biases than the current int32 and the `quantized_bias_type` would be reused for that, a default `TensorType::INT64` would then enter in conflict with the default int32 currently used. We can't use `TensorType::INT32` either as default value as we want old int16 models to use int64 as bias by default. It's why we currently use `TensorType::FLOAT32=0` to mark it as not set though it may not be the cleanest way to do so. We could eventually encapsulate the `quantized_bias_type` with a is set flag, have a `NONE` type or use something else than `TensorType` if it's problematic.

Currently during the `ParseConv2D/FullyConnected/TransposeConv` in `flatbuffer_conversions.cc` the `TensorType` is converted to a `TfLiteType`. We then check in the kernel preparation if `params->quantized_bias_type != kTfLiteFloat32`. We could eventually to this check during the FlatBuffer conversion and use `kTfLiteNoType` when a `TensorType::FLOAT32` is read instead of simply converting it to `kTfLiteFloat32` with `ConvertTensorType`.

The current way of doing in the PR works (we fixed an error in the latest commits) but we can eventually make some changes on how to handle this depending on your preference.

gbaned commented:
@SaoirseARM Can you please resolve conflicts? Thank you!

tucan9389 commented:
@SaoirseARM 
Could you resolve the conflict and PyLint of the failing checks?
Thanks!

SaoirseARM committed:
Merge remote-tracking branch 'origin/master' into toupstream/int6x8_32_accum

SaoirseARM committed:
Fix pylint errors

SaoirseARM commented:
> @SaoirseARM Could you resolve the conflict and PyLint of the failing checks? Thanks!

Hi, 

I have resolved the conflicts and the pylint errors.

Thanks,

Saoirse

gbaned commented:
Hi @tucan9389 Can you please review this PR ? Thank you!

gbaned commented:
Hi @rewu93 Can you please review this PR ? Thank you!

yyoon commented:
@gbaned Hi. We are already in the process of merging this PR. It's taking time to have some internal discussion before merging, because this PR touches the TFLite flatbuffer schema which can have some bigger impact.

No action items for ARM side for now, and hopfully we could merge this soon.

yyoon commented:
cc/ @tucan9389

tucan9389 commented on tensorflow/lite/tools/versioning/runtime_version.cc:
```diff
@@ -124,7 +125,8 @@ std::string FindMinimumRuntimeVersionForOp(tflite::BuiltinOperator op_code,
            {{BuiltinOperator_FULLY_CONNECTED, 8}, "2.3.0"},
            {{BuiltinOperator_FULLY_CONNECTED, 9}, "2.3.0"},
-           {{BuiltinOperator_FULLY_CONNECTED, 10}, "2.11.0"},
+           {{BuiltinOperator_FULLY_CONNECTED, 10}, "2.12.0"},
```
@SaoirseARM 
Sorry for asking after approval. Was there any reason to change the previous version of CONV_2D and FULLY_CONNECTED? If not, I'll change it back myself on our inside process. (Same with line 75 of this file)

tucan9389 commented:
Hi SaoirseARM@
We found several failure tests. Could you check it with the following commands and fix those?

```shell
bazel test //tensorflow/compiler/mlir/lite/quantization/lite:quantize_model_test
```

```shell
bazel test //tensorflow/lite/kernels:transpose_conv_test
```

SaoirseARM committed:
Revert update for previous versions of CONV_2D and TRANPOSE_CONV operators

SaoirseARM commented:
> Hi @SaoirseARM We found several failure tests. Could you check it with the following commands and fix those?
> 
> ```shell
> bazel test //tensorflow/compiler/mlir/lite/quantization/lite:quantize_model_test
> ```
> 
> ```shell
> bazel test //tensorflow/lite/kernels:transpose_conv_test
> ```

Hi @tucan9389, 

Could you provide a little bit more information. I have run both tests on my side and they are both passing. 
Perhaps I am missing something? 

Best regards,

Saoirse

tucan9389 commented:
@SaoirseARM 
Thanks for waiting. 

I could produce the following error message with the command:

```
bazel test //tensorflow/lite/kernels:transpose_conv_test --test_output=all --test_arg=-v=1
```

Part of the error message for the `//tensorflow/lite/kernels:transpose_conv_test`:

```
tensorflow/lite/kernels/transpose_conv_test.cc:1176: Failure
Value of: model.GetDequantizedOutput()
Expected: has 12 elements where
element #0 is approximately 5 (absolute error <= 0.19),
element #1 is approximately 24 (absolute error <= 0.19),
element #2 is approximately 22 (absolute error <= 0.19),
element #3 is approximately 64 (absolute error <= 0.19),
element #4 is approximately 42 (absolute error <= 0.19),
element #5 is approximately 64 (absolute error <= 0.19),
element #6 is approximately 34 (absolute error <= 0.19),
element #7 is approximately 64 (absolute error <= 0.19),
element #8 is approximately 64 (absolute error <= 0.19),
element #9 is approximately 64 (absolute error <= 0.19),
element #10 is approximately 64 (absolute error <= 0.19),
element #11 is approximately 64 (absolute error <= 0.19)
  Actual: { -64, -64, -64, -64, -64, -64, -64, -64, -64, -64, -64, -64 }, whose element #0 doesn't match, which is -69 from 5
[  FAILED  ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias32/0, where GetParam() = ("GenericOptimized", 4-byte object <00-00 00-00>) (0 ms)
[ RUN      ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias32/1
tensorflow/lite/kernels/transpose_conv_test.cc:1176: Failure
Value of: model.GetDequantizedOutput()
Expected: has 12 elements where
element #0 is approximately 5 (absolute error <= 0.19),
element #1 is approximately 24 (absolute error <= 0.19),
element #2 is approximately 22 (absolute error <= 0.19),
element #3 is approximately 64 (absolute error <= 0.19),
element #4 is approximately 42 (absolute error <= 0.19),
element #5 is approximately 64 (absolute error <= 0.19),
element #6 is approximately 34 (absolute error <= 0.19),
element #7 is approximately 64 (absolute error <= 0.19),
element #8 is approximately 64 (absolute error <= 0.19),
element #9 is approximately 64 (absolute error <= 0.19),
element #10 is approximately 64 (absolute error <= 0.19),
element #11 is approximately 64 (absolute error <= 0.19)
  Actual: { -64, -64, -64, -64, -64, -64, -64, -64, -64, -64, -64, -64 }, whose element #0 doesn't match, which is -69 from 5
[  FAILED  ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias32/1, where GetParam() = ("GenericOptimized", 4-byte object <01-00 00-00>) (0 ms)
[ RUN      ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias32/2
[       OK ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias32/2 (0 ms)
[ RUN      ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias32/3
[       OK ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias32/3 (0 ms)
[ RUN      ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias64/0
[       OK ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias64/0 (0 ms)
[ RUN      ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias64/1
[       OK ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias64/1 (0 ms)
[ RUN      ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias64/2
[       OK ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias64/2 (0 ms)
[ RUN      ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias64/3
[       OK ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias64/3 (0 ms)
[----------] 100 tests from TransposeConvOpTest/TransposeConvOpTest (2 ms total)

[----------] Global test environment tear-down
[==========] 100 tests from 1 test suite ran. (2 ms total)
[  PASSED  ] 98 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias32/0, where GetParam() = ("GenericOptimized", 4-byte object <00-00 00-00>)
[  FAILED  ] TransposeConvOpTest/TransposeConvOpTest.SimpleBiasTestQuantizedPerChannel16x8Bias32/1, where GetParam() = ("GenericOptimized", 4-byte object <01-00 00-00>)

 2 FAILED TESTS
```

I also attempted to reproduce the `//tensorflow/compiler/mlir/lite/quantization/lite:quantize_model_test` error in my external bazel environment, but I was unable to do so. I will provide an update once I have more information.

tucan9389 commented on tensorflow/lite/tools/versioning/runtime_version.cc:
```diff
@@ -73,6 +73,7 @@ std::string FindMinimumRuntimeVersionForOp(tflite::BuiltinOperator op_code,
            {{BuiltinOperator_CONV_2D, 5}, "2.4.0"},
            {{BuiltinOperator_CONV_2D, 6}, "2.9.0"},
            {{BuiltinOperator_CONV_2D, 7}, "2.11.0"},
+           {{BuiltinOperator_CONV_2D, 8}, "2.12.0"},
```
@SaoirseARM 
Could you update it to "2.13.0" like `BuiltinOperator_TRANSPOSE_CONV`?

tucan9389 commented on tensorflow/lite/tools/versioning/runtime_version.cc:
```diff
@@ -125,6 +126,7 @@ std::string FindMinimumRuntimeVersionForOp(tflite::BuiltinOperator op_code,
            {{BuiltinOperator_FULLY_CONNECTED, 8}, "2.3.0"},
            {{BuiltinOperator_FULLY_CONNECTED, 9}, "2.3.0"},
            {{BuiltinOperator_FULLY_CONNECTED, 10}, "2.11.0"},
+           {{BuiltinOperator_FULLY_CONNECTED, 11}, "2.12.0"},
```
@SaoirseARM
Could you update it to "2.13.0" like `BuiltinOperator_TRANSPOSE_CONV`?

tucan9389 commented:
@SaoirseARM
I re-ran the kokoro test, and I think you can access and check the detail of the failure in the "`Details`" of the "Py+CPP Test Suite - Ubuntu CPU, Python 3.9 — Internal CI build failed" Check. Please resolve these failures. 

I believe you can also check the error message from `//tensorflow/compiler/mlir/lite/quantization/lite:quantize_model_test` on the 2nd failure([pkg/pip_and_nonpip_tests](https://source.cloud.google.com/results/invocations/33cf66fc-fe3e-49ae-a776-8618618d6009/targets/pkg%2Fpip_and_nonpip_tests)) in the "`Details`".

<details>
  <summary>Error message from `pkg/pip_and_nonpip_tests` failure</summary>
```
tensorflow/compiler/mlir/lite/quantization/lite/quantize_model_test.cc:1221
Value of: model_.operator_codes[0]->version
Expected: is equal to 5
  Actual: 11 (of type int)
NOTE: From /tensorflow/compiler/mlir/lite/quantization/lite/quantize_model_test/test.xml
NOTE: Number of failures for this test: 3.
      Most TF jobs run tests three times to root out flakes.
      Since there were three failures, this is not flaky, and it
      probably caused the Kokoro invocation to fail.
```

</details>

SaoirseARM committed:
Versioning updates and add extra check in quantize_model

tucan9389 commented:
@SaoirseARM 
Hi, thanks for the update. 
Could you again check the error in [the comment](https://github.com/tensorflow/tensorflow/pull/58400#issuecomment-1451494181)? I could reproduce the error message with the following command:

```shell
bazel test //tensorflow/lite/kernels:transpose_conv_test --test_output=all --test_arg=-v=1
```

gbaned commented:
Hi @SaoirseARM Any update on this PR? Please. Thank you!

SaoirseARM committed:
Merge remote-tracking branch 'origin/master' into toupstream/int6x8_32_accum

SaoirseARM committed:
Add fix for quantize_model_test

SaoirseARM committed:
Fix pylint errors

SaoirseARM committed:
Merge remote-tracking branch 'origin/master' into toupstream/int6x8_32_accum

SaoirseARM committed:
Merge remote-tracking branch 'origin/master' into toupstream/int6x8_32_accum

SaoirseARM commented:
Hi, @tucan9389, 

I was wondering if you could review this when you have a chance. 
I unfortunately can't reproduce the failure on the test above.

Thanks,
Saoirse

gbaned commented:
Hi @tucan9389 Can you please review this PR ? Thank you!

gbaned commented:
Hi @tucan9389 Can you please review this PR ? Thank you!

tucan9389 commented:
@SaoirseARM cc. @gbaned, @yyoon 
Sorry for the late reply. Could you merge the 'origin/master' into the current branch? I'll run the internal test, and let's check again whether there is an additional error. (I found no previous error in my local env)

SaoirseARM committed:
Merge remote-tracking branch 'origin/master' into toupstream/int6x8_32_accum

SaoirseARM committed:
Merge remote-tracking branch 'origin/master' into toupstream/int6x8_32_accum

tucan9389 commented:
https://github.com/tensorflow/tensorflow/actions/runs/5014407487/jobs/8988616778
Could you check this error of the Checks and fix it?

SaoirseARM committed:
Merge remote-tracking branch 'origin/master' into toupstream/int6x8_32_accum

tucan9389 commented:
Hi @SaoirseARM,
Could you check [the failure](https://source.cloud.google.com/results/invocations/8f84cfd7-5887-4d07-90d0-d857043ee0e7/targets/tensorflow%2Frel%2Fdocker%2Fgithub_presubmit%2Fpycpp_gpu_py39/log)? 

I think it's from tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py test.

SaoirseARM committed:
Merge remote-tracking branch 'origin/master' into toupstream/int6x8_32_accum

SaoirseARM commented:
> Hi @SaoirseARM, Could you check [the failure](https://source.cloud.google.com/results/invocations/8f84cfd7-5887-4d07-90d0-d857043ee0e7/targets/tensorflow%2Frel%2Fdocker%2Fgithub_presubmit%2Fpycpp_gpu_py39/log)?
> 
> I think it's from tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py test.

Thanks! The tests are now passing but there seems to be other issues with the build. Unsure if these are related to the PR? 

best regards, 
Saoirse

alankelly commented:
Please fix tensorflow/lite/kernels:transpose_conv_test:

```W0802 00:49:23.534827 2118086 thread.cc:2019] --- Thread 7f1003e31d00 (name: main/2118086) stack: ---
W0802 00:49:23.544801 2118086 thread.cc:2019]     @     0x7f100470fd0a  ./third_party/tensorflow/lite/kernels/test_util.h:93 tflite::Quantize<>()
W0802 00:49:23.545504 2118086 thread.cc:2019]     @     0x7f0f2f81ed68  base/process_state.cc:1281 FailureSignalHandler()
W0802 00:49:23.546020 2118086 thread.cc:2019]     @     0x7f1003e7a1c0  __restore_rt
W0802 00:49:23.546037 2118086 thread.cc:2019]     @     0x7f100470f68f  ./third_party/tensorflow/lite/kernels/test_util.h:489 tflite::SingleOpModel::QuantizeAndPopulate<>()
W0802 00:49:23.546051 2118086 thread.cc:2019]     @     0x7f10046a7cef  third_party/tensorflow/lite/kernels/transpose_conv_test.cc:897 tflite::(anonymous namespace)::BaseTransposeConvBiasOpModel<>::SetBias()
W0802 00:49:23.546066 2118086 thread.cc:2019]     @     0x7f10046a778c  third_party/tensorflow/lite/kernels/transpose_conv_test.cc:1169 tflite::(anonymous namespace)::TransposeConvOpTest_SimpleBiasTestQuantizedPerChannel16x8Bias32_Test::TestBody()
W0802 00:49:23.546080 2118086 thread.cc:2019]     @     0x7f0f70d26fa2  third_party/googletest/googletest/src/gtest.cc:2612 testing::internal::HandleSehExceptionsInMethodIfSupported<>()
W0802 00:49:23.546094 2118086 thread.cc:2019]     @     0x7f0f70cf9b2b  third_party/googletest/googletest/src/gtest.cc:2667 testing::internal::HandleExceptionsInMethodIfSupported<>()
W0802 00:49:23.546108 2118086 thread.cc:2019]     @     0x7f0f70cdcf5b  third_party/googletest/googletest/src/gtest.cc:2687 testing::Test::Run()
W0802 00:49:23.546121 2118086 thread.cc:2019]     @     0x7f0f70cdda97  third_party/googletest/googletest/src/gtest.cc:2836 testing::TestInfo::Run()
W0802 00:49:23.546135 2118086 thread.cc:2019]     @     0x7f0f70cde598  third_party/googletest/googletest/src/gtest.cc:3015 testing::TestSuite::Run()
W0802 00:49:23.546209 2118086 thread.cc:2019]     @     0x7f0f70ceba3e  third_party/googletest/googletest/src/gtest.cc:5923 testing::internal::UnitTestImpl::RunAllTests()
W0802 00:49:23.546224 2118086 thread.cc:2019]     @     0x7f0f70d346e2  third_party/googletest/googletest/src/gtest.cc:2612 testing::internal::HandleSehExceptionsInMethodIfSupported<>()
W0802 00:49:23.546238 2118086 thread.cc:2019]     @     0x7f0f70cfd1eb  third_party/googletest/googletest/src/gtest.cc:2667 testing::internal::HandleExceptionsInMethodIfSupported<>()
W0802 00:49:23.546252 2118086 thread.cc:2019]     @     0x7f0f70ceb403  third_party/googletest/googletest/src/gtest.cc:5487 testing::UnitTest::Run()
W0802 00:49:23.546265 2118086 thread.cc:2019]     @     0x7f1004007a79  third_party/googletest/googletest/include/gtest/gtest.h:2332 RUN_ALL_TESTS()
```

Scale is set to zero

SaoirseARM committed:
Merge remote-tracking branch 'origin/master' into toupstream/int6x8_32_accum

SaoirseARM commented:
> Please fix tensorflow/lite/kernels:transpose_conv_test:
> 
> ```
> W0802 00:49:23.544801 2118086 thread.cc:2019]     @     0x7f100470fd0a  ./third_party/tensorflow/lite/kernels/test_util.h:93 tflite::Quantize<>()
> W0802 00:49:23.545504 2118086 thread.cc:2019]     @     0x7f0f2f81ed68  base/process_state.cc:1281 FailureSignalHandler()
> W0802 00:49:23.546020 2118086 thread.cc:2019]     @     0x7f1003e7a1c0  __restore_rt
> W0802 00:49:23.546037 2118086 thread.cc:2019]     @     0x7f100470f68f  ./third_party/tensorflow/lite/kernels/test_util.h:489 tflite::SingleOpModel::QuantizeAndPopulate<>()
> W0802 00:49:23.546051 2118086 thread.cc:2019]     @     0x7f10046a7cef  third_party/tensorflow/lite/kernels/transpose_conv_test.cc:897 tflite::(anonymous namespace)::BaseTransposeConvBiasOpModel<>::SetBias()
> W0802 00:49:23.546066 2118086 thread.cc:2019]     @     0x7f10046a778c  third_party/tensorflow/lite/kernels/transpose_conv_test.cc:1169 tflite::(anonymous namespace)::TransposeConvOpTest_SimpleBiasTestQuantizedPerChannel16x8Bias32_Test::TestBody()
> W0802 00:49:23.546080 2118086 thread.cc:2019]     @     0x7f0f70d26fa2  third_party/googletest/googletest/src/gtest.cc:2612 testing::internal::HandleSehExceptionsInMethodIfSupported<>()
> W0802 00:49:23.546094 2118086 thread.cc:2019]     @     0x7f0f70cf9b2b  third_party/googletest/googletest/src/gtest.cc:2667 testing::internal::HandleExceptionsInMethodIfSupported<>()
> W0802 00:49:23.546108 2118086 thread.cc:2019]     @     0x7f0f70cdcf5b  third_party/googletest/googletest/src/gtest.cc:2687 testing::Test::Run()
> W0802 00:49:23.546121 2118086 thread.cc:2019]     @     0x7f0f70cdda97  third_party/googletest/googletest/src/gtest.cc:2836 testing::TestInfo::Run()
> W0802 00:49:23.546135 2118086 thread.cc:2019]     @     0x7f0f70cde598  third_party/googletest/googletest/src/gtest.cc:3015 testing::TestSuite::Run()
> W0802 00:49:23.546209 2118086 thread.cc:2019]     @     0x7f0f70ceba3e  third_party/googletest/googletest/src/gtest.cc:5923 testing::internal::UnitTestImpl::RunAllTests()
> W0802 00:49:23.546224 2118086 thread.cc:2019]     @     0x7f0f70d346e2  third_party/googletest/googletest/src/gtest.cc:2612 testing::internal::HandleSehExceptionsInMethodIfSupported<>()
> W0802 00:49:23.546238 2118086 thread.cc:2019]     @     0x7f0f70cfd1eb  third_party/googletest/googletest/src/gtest.cc:2667 testing::internal::HandleExceptionsInMethodIfSupported<>()
> W0802 00:49:23.546252 2118086 thread.cc:2019]     @     0x7f0f70ceb403  third_party/googletest/googletest/src/gtest.cc:5487 testing::UnitTest::Run()
> W0802 00:49:23.546265 2118086 thread.cc:2019]     @     0x7f1004007a79  third_party/googletest/googletest/include/gtest/gtest.h:2332 RUN_ALL_TESTS()
> ```
> 
> Scale is set to zero



> Please fix tensorflow/lite/kernels:transpose_conv_test:
> 
> ```
> W0802 00:49:23.544801 2118086 thread.cc:2019]     @     0x7f100470fd0a  ./third_party/tensorflow/lite/kernels/test_util.h:93 tflite::Quantize<>()
> W0802 00:49:23.545504 2118086 thread.cc:2019]     @     0x7f0f2f81ed68  base/process_state.cc:1281 FailureSignalHandler()
> W0802 00:49:23.546020 2118086 thread.cc:2019]     @     0x7f1003e7a1c0  __restore_rt
> W0802 00:49:23.546037 2118086 thread.cc:2019]     @     0x7f100470f68f  ./third_party/tensorflow/lite/kernels/test_util.h:489 tflite::SingleOpModel::QuantizeAndPopulate<>()
> W0802 00:49:23.546051 2118086 thread.cc:2019]     @     0x7f10046a7cef  third_party/tensorflow/lite/kernels/transpose_conv_test.cc:897 tflite::(anonymous namespace)::BaseTransposeConvBiasOpModel<>::SetBias()
> W0802 00:49:23.546066 2118086 thread.cc:2019]     @     0x7f10046a778c  third_party/tensorflow/lite/kernels/transpose_conv_test.cc:1169 tflite::(anonymous namespace)::TransposeConvOpTest_SimpleBiasTestQuantizedPerChannel16x8Bias32_Test::TestBody()
> W0802 00:49:23.546080 2118086 thread.cc:2019]     @     0x7f0f70d26fa2  third_party/googletest/googletest/src/gtest.cc:2612 testing::internal::HandleSehExceptionsInMethodIfSupported<>()
> W0802 00:49:23.546094 2118086 thread.cc:2019]     @     0x7f0f70cf9b2b  third_party/googletest/googletest/src/gtest.cc:2667 testing::internal::HandleExceptionsInMethodIfSupported<>()
> W0802 00:49:23.546108 2118086 thread.cc:2019]     @     0x7f0f70cdcf5b  third_party/googletest/googletest/src/gtest.cc:2687 testing::Test::Run()
> W0802 00:49:23.546121 2118086 thread.cc:2019]     @     0x7f0f70cdda97  third_party/googletest/googletest/src/gtest.cc:2836 testing::TestInfo::Run()
> W0802 00:49:23.546135 2118086 thread.cc:2019]     @     0x7f0f70cde598  third_party/googletest/googletest/src/gtest.cc:3015 testing::TestSuite::Run()
> W0802 00:49:23.546209 2118086 thread.cc:2019]     @     0x7f0f70ceba3e  third_party/googletest/googletest/src/gtest.cc:5923 testing::internal::UnitTestImpl::RunAllTests()
> W0802 00:49:23.546224 2118086 thread.cc:2019]     @     0x7f0f70d346e2  third_party/googletest/googletest/src/gtest.cc:2612 testing::internal::HandleSehExceptionsInMethodIfSupported<>()
> W0802 00:49:23.546238 2118086 thread.cc:2019]     @     0x7f0f70cfd1eb  third_party/googletest/googletest/src/gtest.cc:2667 testing::internal::HandleExceptionsInMethodIfSupported<>()
> W0802 00:49:23.546252 2118086 thread.cc:2019]     @     0x7f0f70ceb403  third_party/googletest/googletest/src/gtest.cc:5487 testing::UnitTest::Run()
> W0802 00:49:23.546265 2118086 thread.cc:2019]     @     0x7f1004007a79  third_party/googletest/googletest/include/gtest/gtest.h:2332 RUN_ALL_TESTS()
> ```
> 
> Scale is set to zero

Hi @alankelly , 

Could you elaborate on this failure a bit more, I can see that the scale is set: https://github.com/tensorflow/tensorflow/pull/58400/commits/ca1ef4e3d419eaf73b8f0a4be0dbf720011e10f8#diff-aedb3c8ef1dcf547c874f8da6d0be81a19f93be0a5f7da53209a4068cb18bedbR907 
but perhaps this refers to something else?

Thanks,
Saoirse

alankelly commented:
Hi Saoirse,

The issue can be recreated using sanitizers: `bazel run  -c dbg --config=asan tensorflow/lite/kernels:transpose_conv_test` I'm not sure that whether you use bazel or not or whether sanitizers will work out of the box, but adding a build flag `-fsanitize=address` should probably work. 

It looks like the test is silently failing without sanitizers, the bias scale is always zero. When you set a model up with per channel quantized filter, the bias is also per channel quantized:

```    if (filter.per_channel_quantization) {
        // per channel quantization.
        std::vector<float> bias_scale(
            filter.per_channel_quantization_scales.size());
        std::vector<int64_t> bias_zero_points(
            filter.per_channel_quantization_scales.size());
        for (size_t i = 0; i < filter.per_channel_quantization_scales.size();
             ++i) {
          bias_scale[i] =
              input.scale * filter.per_channel_quantization_scales[i];
          bias_zero_points[i] = 0;
        }
        TensorData bias{bias_type,
                        {bias_size},
                        /*min=*/0,
                        /*max=*/0,
                        /*scale=*/0,
                        /*zero_point=*/0,
                        true,
                        /*per_channel_quantization_scales=*/bias_scale,
                        /*per_channel_quantization_offsets=*/bias_zero_points,
                        /*channel_index==*/0};
        bias_ = AddInput(bias);```

The call to ` model.SetBias({3, -2});` in the new test calls `QuantizeAndPopulate<int32_t>(bias_, data)` which calls `void QuantizeAndPopulate(int index, const std::vector<float>& data) ` and then `auto q = Quantize<T>(data, t->params.scale, t->params.zero_point, t->type);`
 
t->params.scale is used to quantize the bias tensor which hasn't been set, so the bias value is divided by zero:

```  q.push_back(static_cast<T>(std::max<float>(                                                                                             
         min, std::min<float>(max, std::round(zero_point + (f / scale))))));     `` `

This looks like a test issue. However, I don't see where the per channel bias scale is later used. Are you sure it's needed? Either fix the test to properly quantize the bias using the per-channel scales or if the bias does not use per-channel scales, then just scale it using a per-tensor value. Some better testing is probably needed as this issue should have been caught by the test and not the sanitizer.

gbaned commented:
Hi @SaoirseARM Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SaoirseARM Any update on this PR? Please. Thank you!

SaoirseARM committed:
Merge remote-tracking branch 'origin/master' into toupstream/int6x8_32_accum

SaoirseARM committed:
Update Transpose_conv_test

SaoirseARM committed:
Update versions

SaoirseARM committed:
Merge branch 'tensorflow:master' into toupstream/int6x8_32_accum

alankelly commented on tensorflow/lite/kernels/conv.cc:
```diff
@@ -396,6 +394,20 @@ TfLiteStatus Prepare(KernelType kernel_type, TfLiteContext* context,
+    TF_LITE_ENSURE_EQ(context, output->params.zero_point, 0);
+
+    // Check we have quantized_bias_type is either kTfLiteInt64 or kTfLiteInt32.
+    if (params->quantized_bias_type != kTfLiteFloat32) {
```
Comment from internal reviewer: "Why not kTfLiteNoType ? Can quantized_bias_type be Float32?"

Tessil replied:
See the https://github.com/tensorflow/tensorflow/pull/58400#issuecomment-1367898150 comment for details regarding this choice. The main problem is that the FlatBuffer scheme TensorType doesn't have a NONE type which could be used as default value for older TFLite models.

alankelly commented on tensorflow/lite/kernels/fully_connected.cc:
```diff
@@ -1135,21 +1146,22 @@ void FullyConnectedPerChannelInt16(const OpData* data,
         GetTensorData<int16_t>(input), GetTensorShape(filter),
         GetTensorData<int8_t>(filter), GetTensorShape(bias),
-        GetTensorData<int64_t>(bias), GetTensorShape(output),
+        GetTensorData<int32_t>(bias), GetTensorShape(output),
```
Comment from internal reviewer: "why do we use std::int32_t in the similar code in conv.cc?"

alankelly commented:
Don't waste your time synching this PR, I have patched a version of it internally and I will handle updates. This should land soon.

SaoirseARM commented:
> Don't waste your time synching this PR, I have patched a version of it internally and I will handle updates. This should land soon.

ok, thank you very much!
- Saoirse

alankelly commented on tensorflow/lite/kernels/fully_connected.cc:
```diff
@@ -1329,6 +1337,11 @@ TfLiteStatus EvalQuantized(TfLiteContext* context, TfLiteNode* node,
+                           : FullyConnectedInt16<kernel_type>(
+                                 data, input, filter, bias, output);
           }
         } else if (kernel_type == kReference) {
```
This has broken an internal test. Could you please explain the change in logic here?

alankelly commented:
This has broken an internal test. Please see https://github.com/tensorflow/tensorflow/pull/58400/files#r1325560610

SaoirseARM commented:
> This has broken an internal test. Please see https://github.com/tensorflow/tensorflow/pull/58400/files#r1325560610



Hi @alankelly,

This change to the logic was changed to ensure even if there is no bias, we default to int64_t and not int32_t.

Is there a way of pointing us to the failing test?

Thanks,

Saoirse
